README.md
# meteor-ping

[![Build Status](https://travis-ci.org/meonkeys/meteor-ping.svg?branch=master)](https://travis-ci.org/meonkeys/meteor-ping)

Ping a [Meteor](https://www.meteor.com) app. This "DDP ping" only succeeds if a Meteor-specific test (subscribing to a collection) passes.

Timing is done internally to adjust for Node.js startup time.

This code is used heavily and frequently for pinging many Meteor apps in production.

## Synopsis

### Command-line use

```bash
npm install -g meteor-ping
meteor-ping --host www.meteor.com --port 443
```

Upon failure, a message is printed to standard error and the program exits with an error code.

Upon success, the milliseconds elapsed for the ping is printed to standard output.

### API

```bash
npm install --save meteor-ping
```

The only public function is `ping()`, which takes one argument, a [standard Node.js-style callback](http://fredkschott.com/post/2014/03/understanding-error-first-callbacks-in-node-js/). The second argument to the callback is an object with one key: `elapsedTimeInMs`. Example usage:

```javascript
'use strict';

var MeteorPing = require('meteor-ping');

// all args are optional, here are the defaults
var localApp = new MeteorPing({
  host: 'localhost',
  port: 3000,
  ssl: false,
  timeout: 10 * 1000, // 10 seconds
});

localApp.ping(function(error, result) {
  if (error) {
    console.error(error);
    process.exit(1);
  } else {
    console.log('done. Milliseconds elapsed: ' + result.elapsedTimeInMs);
  }
});
```

### MeteorPing constructor arguments

All arguments are optional.

1. `host` - domain name of Meteor server to ping. Must be a valid IP address or hostname.
    * Default `localhost`.
1. `port` - port of Meteor server to ping. Must be a valid port number.
    * Default `3000`.
1. `ssl` - whether to use SSL for the connection. Boolean. Automatically `true` if `port` is `443` (no idea why).
    * Default `false`.
1. `connectTimeout` - milliseconds before timing out the DDP `connect()` call. Must be a number greater than or equal to 1.
    * Default `5000`.
1. `subscribeTimeout` - milliseconds before timing out the DDP `subscribe()` call. Must be a number greater than or equal to 1.
    * Default `5000`.
1. `collection` - collection to use for `subscribe()` call. Must be a string. Consider using a database-backed collection if you have one--this provides a full round-trip ping.
    * Default `meteor.loginServiceConfiguration`.

## Description

Meteor apps deliver static content and layout. A websocket or long-polling connection is used to deliver data. An app may return a HTTP 200 OK even if data cannot be retrieved. It is helpful to have a smarter check to ensure the static content **and** data are available.

The pinged app must use the [Accounts system](https://docs.meteor.com/#/full/accounts_api) or must override the `collection` constructor argument.

## Compatibility

This package has been tested with Node.js version 4.6.1.

## Copyright, License

Copyright (C) 2016 Adam Monsen

MIT license.


CHANGELOG.md
## 1.0.3 - 2017-01-03

* Prevent double callback bug in ddp `subscribe()`.

## 1.0.2 - 2016-10-27

* Fix bug: actually use `subscribeTimeout` instead of just `connectTimeout` for both `subscribe()` and `connect()` timeouts.

## 1.0.1 - 2016-09-09

* Prevent double callback bug in ddp `connect()`.

## 1.0.0 - 2016-09-07

* Improve timeout: use consistent [error-first callback](http://fredkschott.com/post/2014/03/understanding-error-first-callbacks-in-node-js/) behavior.
* Add `connectTimeout` and `subscribeTimeout` params on the command line and API.
* Drop support for Node.js < v4.x.x. `let`, `const`, and string templates ahoy!

## 0.2.3 - 2016-09-06

* Support arbitrary collection for `subscribe()` call.
* Upgrade ddp to latest (v0.12.0).

## 0.2.2 - 2016-07-28

* Refactor into a module and a library, and support use of either independently.

## 0.1.2 - 2016-03-17

* First release!


test/basic.js
'use strict';

var assert = require('chai').assert;
var MeteorPing = require('../lib/index.js');

describe('Setup', function() {
  describe('init', function() {
    it('should return a properly initialized MeteorPing object', function() {
      var m = new MeteorPing();
      assert.equal(m.args.host, 'localhost');
    });

    it('rejects unexpected constructor args', function() {
      function testThrow() {
        new MeteorPing({password: 'blah'});
      }
      assert.throws(testThrow,
        'unexpected arg(s) passed to MeteorPing constructor');
    });
  });
});

// - pseudo-e2e test with a 1ms timeout
describe('Ping', function() {
  describe('simple attempt', function() {
    it('no errors on ping attempt and calls callback', function(done) {
      var m = new MeteorPing({connectTimeout: 1});
      m.ping(function() {
        // discard error & result
        done();
      });
    });
  });
});

// TODO: add e2e test, with minimal local meteor app (ideally several... one
// with slow publications)


main.js
#!/usr/bin/env node

'use strict';

var Promise = require('promise-polyfill');
var DDPClient = require('ddp');
var debugPkg = require('debug');
var yargs = require('yargs');

var argv = yargs
  .usage('Usage: $0 [options]')
  .boolean(['noRetry', 'ssl'])
  .default({
    noRetry: false,
    retryInterval: 2000, // milliseconds
    maxRetries: 60,
    host: 'localhost',
    port: 3000,
    ssl: false
  })
  .help('h')
  .alias('h', 'help')
  .argv;

var debug = debugPkg('upcheck');
debug('start');

var ddpclient = new DDPClient({
  host: argv.host,
  port: argv.port,
  ssl: argv.ssl,
  autoReconnect: false, // only re-connects, doesn't re-try first connect
  maintainCollections: true,
  ddpVersion: '1',
});

ddpclient.on('message', function (msg) {
  debug('DDP message: ' + msg);
});

ddpclient.on('socket-close', function(code, message) {
  debug('Socket closed. code=' + code + ' message=' + message);
});

ddpclient.on('socket-error', function(error) {
  debug('Socket error: ' + error.message);
});

var retries = 0;
function connect() {
  return new Promise(function(resolve, reject) {
    ddpclient.connect(function(error, wasReconnect){
      if (error) {
        ddpclient.close();
        if (!argv.noRetry && (retries < argv.maxRetries)) {
          debug('retry ' + (retries + 1) + ' of ' + argv.maxRetries);
          retries++;
          setTimeout(function() {
            connect().then(function() {
              resolve();
            }); // should we have a catch here?
          }, argv.retryInterval);
        } else {
          reject(new Error('connection to ' + argv.host + ':' + argv.port + ' failed and exceeded maxRetries'));
          return;
        }
      } else {
        if (wasReconnect) {
          // we should never get here when autoReconnect is false
          reject('Unexpected condition: connection reestablished.');
          return;
        }
        resolve();
      }
    });
  });
}

function subscribe() {
  return new Promise(function(resolve, reject) {
    ddpclient.subscribe(
      'meteor.loginServiceConfiguration',
      [],
      function () {
        resolve();
      }
    );
  });
}

connect()
  .then(subscribe)
  .then(function() {
    console.log(argv.host + ':' + argv.port + ' is ready.');
    ddpclient.close();
    debug('end');
    process.exit(0);
  })
  .catch(function(reason) {
    // specific error provided by ddpclient socket-error emitter
    if (reason) {
      console.error(reason);
    } else {
      console.error('Abort.');
    }
    debug('end, with errors');
    process.exit(1);
  });

// vim: ft=javascript


lib/index.js
'use strict';

const _ = require('lodash');
const async = require('async');
const check = require('check-types');
const DDPClient = require('ddp');
const debugPkg = require('debug');

const ddpDebug = debugPkg('ddp');
const pkgname = 'meteor-ping';
const debug = debugPkg(pkgname);

const defaultConstructorArgs = {
  host: 'localhost',
  port: 3000,
  ssl: false,
  connectTimeout: 5 * 1000,
  subscribeTimeout: 5 * 1000,
  collection: 'meteor.loginServiceConfiguration',
};

function MeteorPing(args) {
  check.assert.maybe.object(args,
    'MeteorPing constructor args must be null, undefined, or a plain old object');
  const empty = _.omit(args, ['host', 'port', 'ssl', 'connectTimeout', 'subscribeTimeout', 'collection']);
  check.assert.emptyObject(empty, 'unexpected arg(s) passed to MeteorPing constructor');
  this.args = _.defaults(args, defaultConstructorArgs);
  // https://stackoverflow.com/a/106223/156060 has some better ideas for hostname checking
  check.assert.string(this.args.host,
    'host string passed to MeteorPing constructor must be a valid IP address or hostname');
  check.assert.inRange(this.args.port, 0, 65535,
    'port passed to MeteorPing constructor must be a valid port number');
  check.assert.boolean(this.args.ssl,
    'invalid ssl boolean passed to MeteorPing constructor');
  check.assert.greaterOrEqual(this.args.connectTimeout, 1,
    'connectTimeout milliseconds passed to MeteorPing constructor must be a number >= 1');
  check.assert.greaterOrEqual(this.args.subscribeTimeout, 1,
    'subscribeTimeout milliseconds passed to MeteorPing constructor must be a number >= 1');
  check.assert.string(this.args.collection,
    'collection parameter to MeteorPing constructor must be a string');
}

function setupDdp(data, callback) {
  debug('setupDdp');
  data.ddpclient = new DDPClient({
    host: data.host,
    port: data.port,
    ssl: data.ssl,
    autoReconnect: false, // only affects re-connects (won't enable re-try of initial connect)
    maintainCollections: true,
    ddpVersion: '1',
  });

  data.ddpclient.on('message', function (msg) {
    ddpDebug('DDP message: ' + msg);
  });

  data.ddpclient.on('socket-close', function(code, message) {
    ddpDebug('Socket closed. code=' + code + ' message=' + message);
  });

  data.ddpclient.on('socket-error', function(error) {
    ddpDebug('Socket error: ' + error.message);
  });

  callback(null, data);
}

function connect(data, callback) {
  debug('connect to ' + data.host + ':' + data.port + ', timeout ' + data.connectTimeout + 'ms');

  let callbackFired = false;
  const timeout = setTimeout(function() {
    if (!callbackFired) {
      callbackFired = true;
      const error = `${pkgname} connect to ${data.host}:${data.port} timed out after ${data.connectTimeout}ms`;
      callback(error, data);
    }
  }, data.connectTimeout);

  data.ddpclient.connect(function(error, wasReconnect){
    if (callbackFired) {
      return;
    }
    clearTimeout(timeout); // abort timeout callback (above)

    if (wasReconnect) {
      // we should never get here when autoReconnect is false
      throw `${pkgname} Unexpected condition: connection reestablished to ${data.host}:${data.port}`;
    }

    debug('Connected to ' + data.host);

    callback(error, data);
  });
}

function subscribe(data, callback) {
  debug('subscribe to ' + data.collection);

  let callbackFired = false;
  const timeout = setTimeout(function() {
    if (!callbackFired) {
      callbackFired = true;
      const error = `${pkgname} subscribe to ${data.collection} timed out after ${data.subscribeTimeout}ms for ${data.host}:${data.port}`;
      callback(error, data);
    }
  }, data.subscribeTimeout);

  data.ddpclient.subscribe(
    data.collection,
    [],
    function (error) {
      if (callbackFired) {
        return;
      }
      clearTimeout(timeout); // abort timeout callback (above)
      callback(error, data);
    }
  );
}

function close(data, callback) {
  debug('close');
  if (data.ddpclient.socket) {
    data.ddpclient.close();
  }
  callback(null, data);
}

MeteorPing.prototype.ping = function(callback) {
  debug('start');
  this.args.startTime = new Date().getTime();
  async.waterfall([
    async.apply(setupDdp, this.args),
    connect,
    subscribe,
    close,
  ], function(error, data) {
    debug('in waterfall callback');
    const elapsed = new Date().getTime() - data.startTime;
    const result = {elapsedTimeInMs: elapsed};
    callback(error, result);
  });
};

module.exports = MeteorPing;
module.exports.defaultConstructorArgs = defaultConstructorArgs;
